src/aro/Compilation.zig

```diff
@@ -1216,103 +1231,176 @@ fn addSourceFromPathExtra(comp: *Compilation, path: []const u8, kind: Source.Kin
     return comp.addSourceFromOwnedBuffer(contents, path, kind);
 }
 
-pub const IncludeDirIterator = struct {
-    comp: *const Compilation,
-    cwd_source_id: ?Source.Id,
-    include_dirs_idx: usize = 0,
-    sys_include_dirs_idx: usize = 0,
-    tried_ms_cwd: bool = false,
+pub fn hasInclude(
+    comp: *Compilation,
+    filename: []const u8,
+    includer_token_source: Source.Id,
+    /// angle bracket vs quotes
+    include_type: IncludeType,
+    /// __has_include vs __has_include_next
+    which: WhichInclude,
+) Compilation.Error!bool {
+    if (try FindInclude.run(comp, filename, switch (which) {
+        .next => .{ .only_search_after_dir = comp.getSource(includer_token_source).path },
+        .first => switch (include_type) {
+            .quotes => .{ .allow_same_dir = comp.getSource(includer_token_source).path },
+            .angle_brackets => .only_search,
+        },
+    })) |_| {
+        return true;
+    } else {
+        return false;
+    }
+}
 
-    const FoundSource = struct {
-        path: []const u8,
+const FindInclude = struct {
+    comp: *Compilation,
+    include_path: []const u8,
+    /// We won't actually consider any include directories until after this directory.
+    wait_for: ?[]const u8,
+
+    const Result = struct {
+        source: Source.Id,
         kind: Source.Kind,
+        used_ms_search_rule: bool,
     };
 
-    fn next(self: *IncludeDirIterator) ?FoundSource {
-        if (self.cwd_source_id) |source_id| {
-            self.cwd_source_id = null;
-            const path = self.comp.getSource(source_id).path;
-            return .{ .path = std.fs.path.dirname(path) orelse ".", .kind = .user };
+    fn run(
+        comp: *Compilation,
+        include_path: []const u8,
+        search_strat: union(enum) {
+            allow_same_dir: []const u8,
+            only_search,
+            only_search_after_dir: []const u8,
+        },
+    ) Allocator.Error!?Result {
+        var find: FindInclude = .{
+            .comp = comp,
+            .include_path = include_path,
+            .wait_for = null,
+        };
+
+        if (std.fs.path.isAbsolute(include_path)) {
+            switch (search_strat) {
+                .allow_same_dir, .only_search => {},
+                .only_search_after_dir => return null,
+            }
+            // TODO: classify absolute file as belonging to system includes or not?
```

Comment:
Based on line directives and https://github.com/llvm/llvm-project/blob/f4db14229cd975822c41376afda9d56a29f9396c/clang/lib/Lex/HeaderSearch.cpp#L926 they are user includes.